### ISSUE TITLE
dependency() can leak gcc-isms to rustc

### ISSUE THREAD

**Issue by linkmauve** [mesonbuild/meson#9775]
**Describe the bug**
When passing `dependency()` to a Rust target, if the .pc file includes a non-system header, meson passes it to the `rustc` invocation, making it fail.  Here is an example using valgrind.pc:
```ninja
[1/1] Compiling Rust source ../test.rs
FAILED: libtestcase.so 
rustc -C linker=cc --color=always --crate-type dylib -g -C relocation-model=pic -I/usr/include/valgrind --crate-name testcase --emit dep-info=testcase.d --emit link -o libtestcase.so -C prefer-dynamic ../test.rs
error: Unrecognized option: 'I'

ninja: build stopped: subcommand failed.
```

**To Reproduce**
```meson
project('testcase', 'rust')
shared_library('testcase', 'test.rs', dependencies: dependency('valgrind'))
```
`test.rs` can contain anything, the issue happens before that.

**Expected behavior**
meson would not pass the `-I` to `rustc`, nor any other argument this compiler doesn’t know, and the compilation would continue properly.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? Native build.
* what operating system: up to date ArchLinux
* what Python version are you using: 3.10.1
* what `meson --version`: 0.60.3
* what `ninja --version` if it's a Ninja build: 1.10.2

**Comment by eli-schwartz** [comment]
How should meson know whether any given argument is vital to operation?

I think what you actually want to do is divide the dependency object in two:
- pass to shared_library `dependencies: valgrind_dep.partial_dependency(link_args: true, links: true)`
- pass to bindgen `valgrind_dep.partial_dependency(compile_args: true, includes: true)`

... except the unstable_rust module cannot actually process a dependency object? It seems to solely assume you are bindgen'ing a header that is in your project sources.

**Comment by dcbaker** [comment]
Yeah, right now the bindgen helper in the rust module assumes you're working with a local header (since that's what my use case was when I added it).

We really should figure out the includes though to rust targets automatically, but we really need to do some refactoring of the way dependencies work to fix that I think.

**Comment by bonzini** [comment]
Are there any compile_args that rustc should receive?

**Comment by dcbaker** [comment]
Likely yes, but not that we can handle. Our handling of dependencies is currently very C-family centric. We assume that compile_args are C-like in basically every case.

**Comment by bonzini** [comment]
That's what I thought, so it's easier to just ignore them altogether in rustc command lines.

**Comment by dcbaker** [comment]
Likely for now, yes. We should mark that ignoring them is a work around
